agent/agent_endpoint_test.go

```diff
@@ -6202,6 +6202,104 @@ func TestAgentConnectCALeafCert_goodNotLocal(t *testing.T) {
 	}
 }
 
+func TestAgentConnectCALeafCert_nonBlockingQuery_after_blockingQuery_shouldNotBlock(t *testing.T) {
+	// see: https://github.com/hashicorp/consul/issues/12048
+
+	runStep := func(t *testing.T, name string, fn func(t *testing.T)) {
+		t.Helper()
+		if !t.Run(name, fn) {
+			t.FailNow()
+		}
+	}
+
+	if testing.Short() {
+		t.Skip("too slow for testing.Short")
+	}
+
+	t.Parallel()
+
+	a := NewTestAgent(t, "")
+	testrpc.WaitForTestAgent(t, a.RPC, "dc1")
+	testrpc.WaitForActiveCARoot(t, a.RPC, "dc1", nil)
+
+	{
+		// Register a local service
+		args := &structs.ServiceDefinition{
+			ID:      "foo",
+			Name:    "test",
+			Address: "127.0.0.1",
+			Port:    8000,
+			Check: structs.CheckType{
+				TTL: 15 * time.Second,
+			},
+		}
+		req, err := http.NewRequest("PUT", "/v1/agent/service/register", jsonReader(args))
+		require.NoError(t, err)
+		resp := httptest.NewRecorder()
+		a.srv.h.ServeHTTP(resp, req)
+		if !assert.Equal(t, 200, resp.Code) {
+			t.Log("Body: ", resp.Body.String())
+		}
+	}
+
+	var (
+		serialNumber string
+		index        string
+		issued       structs.IssuedCert
+	)
+	runStep(t, "do initial non-blocking query", func(t *testing.T) {
+		req, err := http.NewRequest("GET", "/v1/agent/connect/ca/leaf/test", nil)
+		require.NoError(t, err)
+		resp := httptest.NewRecorder()
+		a.srv.h.ServeHTTP(resp, req)
+
+		dec := json.NewDecoder(resp.Body)
+		require.NoError(t, dec.Decode(&issued))
+		serialNumber = issued.SerialNumber
+
+		require.Equal(t, "MISS", resp.Header().Get("X-Cache"),
+			"for the leaf cert cache type these are always MISS")
+		index = resp.Header().Get("X-Consul-Index")
+	})
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go func() {
+		// launch goroutine for blocking query
+		resp := httptest.NewRecorder()
+		req, err := http.NewRequestWithContext(ctx, "GET", "/v1/agent/connect/ca/leaf/test?index="+index, nil)
+		if err != nil {
+			t.Errorf("error: %v", err)
+			return
+		}
+		a.srv.h.ServeHTTP(resp, req)
+	}()
+
+	// We just need to ensure that the above blocking query is in-flight before
+	// the next step, so do a little sleep.
+	time.Sleep(50 * time.Millisecond)
+
+	runStep(t, "do a non-blocking query that should not block", func(t *testing.T) {
```

Comment:
For those of us less familiar with this code-path, does this test assume that the waiter won't have been cleared because leaf certificates are generated infrequently? Would it be worth calling out in a comment?

Reply:
I could add it here I guess. A lot of the API-side endpoint tests that test something about caching implicitly rely upon caching behavior where the cache takes longer than the execution of a test suite to prune.

So in this case the original non-blocking query implicitly populated the cache for this leaf entry. Also the initial non-blocking internal query used by the still-running goroutine up from line 6270 also refreshed that entry again. This new non-blocking query will see that the cache entry is present and usable/valid and return without blocking for the life of the above blocking query (as is the errant behavior we're fixing).